typing/typedecl.ml

```diff
@@ -652,60 +651,63 @@ let check_well_founded env loc path to_check ty =
+        visited := TypeMap.add ty parents !visited;
+        (false, parents)
       with Not_found ->
+        visited := TypeMap.add ty parents !visited;
         (false, parents)
     in
     if fini then () else
     let rec_ok =
       match get_desc ty with
-        Tconstr(p,_,_) ->
+      | Tconstr(p,_,_) ->
           !Clflags.recursive_types && Ctype.is_contractive env p
       | Tobject _ | Tvariant _ -> true
       | _ -> !Clflags.recursive_types
     in
-    let visited' = TypeMap.add ty parents !visited in
-    let arg_exn =
-      try
-        visited := visited';
-        let parents =
-          if rec_ok then TypeSet.empty else TypeSet.add ty parents in
-        Btype.iter_type_expr (check ty0 parents) ty;
-        None
-      with e ->
-        visited := visited'; Some e
-    in
+    if rec_ok then () else
+    let parents = TypeSet.add ty parents in
     match get_desc ty with
-    | Tconstr(p, _, _) when arg_exn <> None || to_check p ->
-        if to_check p then Option.iter raise arg_exn
-        else Btype.iter_type_expr (check ty0 TypeSet.empty) ty;
-        begin try
-          let ty' = Ctype.try_expand_once_opt env ty in
-          let ty0 = if TypeSet.is_empty parents then ty else ty0 in
-          check ty0 (TypeSet.add ty parents) ty'
-        with
-          Ctype.Cannot_expand -> Option.iter raise arg_exn
+    | Tconstr(p, tyl, _) ->
+        let to_check = to_check p in
+        if to_check then List.iter (check parents) tyl;
+        begin match Ctype.try_expand_once_opt env ty with
+        | ty' -> check parents ty'
+        | exception Ctype.Cannot_expand ->
+            if not to_check then List.iter (check parents) tyl
         end
-    | _ -> Option.iter raise arg_exn
+    | _ ->
+        Btype.iter_type_expr (check parents) ty
   in
   let snap = Btype.snapshot () in
-  try Ctype.wrap_trace_gadt_instances env (check ty TypeSet.empty) ty
+  try Ctype.wrap_trace_gadt_instances env (check TypeSet.empty) ty0
   with Ctype.Escape _ ->
     (* Will be detected by check_recursion *)
     Btype.backtrack snap
 
 let check_well_founded_manifest env loc path decl =
   if decl.type_manifest = None then () else
   let args = List.map (fun _ -> Ctype.newvar()) decl.type_params in
-  check_well_founded env loc path (Path.same path) (Ctype.newconstr path args)
+  let visited = ref TypeMap.empty in
+  check_well_founded env loc path (Path.same path) visited
+    (Ctype.newconstr path args)
 
 let check_well_founded_decl env loc path decl to_check =
   let open Btype in
+  let visited = ref TypeMap.empty in
+  let checked = ref TypeSet.empty in
   let it =
-    {type_iterators with
-     it_type_expr = (fun _ -> check_well_founded env loc path to_check)} in
+    {type_iterators with it_type_expr =
+     (fun self ty ->
+       if TypeSet.mem ty !checked then () else begin
+         check_well_founded env loc path to_check visited ty;
+         checked := TypeSet.add ty !checked;
+         self.it_type_expr self ty
```

Comment:
I don't think that this call `self.it_type_expr self ty` is correct. The intent is to call the "super" iterator that will recursively traverse all type sub-expressions, but I think that in practice it will call exactly the same function recursively, which will stop as it already visited the node.

See <https://github.com/ocaml/ocaml/blob/efd714b73b41df468f53360a067b54370ab0f98e/typing/typemod.ml#L263> for an example of "super" recursion with type_iterators.

Do you have an idea of a testsuite example that would distinguish between the incorrect and the incorrect behavior here? You mentioned in your discussion that you really want to check non-root local cycles in this code. Is there a known example where this must be done?

Reply:
Well spotted. Indeed the call should not be to `self.it_type_expr` but `self.it_do_type_expr` (or use `super` as you mentioned, which does the same thing).
However, this looks more complex than that, as just doing it that way causes uncaught exceptions in the testsuite.
Will look into that.

Reply:
Actually, the errors were just caused by conflicts with installed files. I wish the testsuite ignored those...

I have committed a fixed version. The only difference is a change in error message in `typing-misc/constraints.ml`.
The change looks like a correct consequence, but this is not really a counter-example, as the code would have been rejected on regularity grounds anyway.

Reply:
BTW, the difference in principal mode comes from two things: one is the different handling of `as` in normal and principal mode, with a loss of sharing in principal, and the second is that, compared to the original approach where the deep recursion was done inside `check_well_founded`, and as result checked the whole expanded graph, this version only check all nodes in the unexpanded graph. So the extra complexity of mixing the checks may be worth it after all.

Reply:
Do we understand why we want to check for local cycles after expansions? Intuitively, if an abbreviation is not mutually-recursive (`to_check`), then we can already assume that its definition does not contain local cycles, and there should be no need to check anything about its expansion, only the arguments. A "neighbor" (`to_check`) declaration may introduce cycles, but it will also be checked in turn by `check_well_founded`. So unless I am missing something, the issue you must have in mind is an abbreviation that does not contain any cycle, but whose expansion introduces a cycle. Is that right? Do we have an example?